File: artifacts/tulipindicators/decompose/src/aroon.c
int ti_aroon_start(const double *options)
{
  return (int) options[0];
}


/*** DEPENDENCIES:
***/


// hint:  ['adown_idx_ref is a mutable refrence to unsigned int', 'maxi_ref is a mutable refrence to int', 'mini_ref is a mutable refrence to int', 'max_ref is a mutable refrence to double', 'min_ref is a mutable refrence to double', 'bar_ref is a mutable refrence to double', 'j_ref is a mutable refrence to int']
void helper_ti_aroon_1(unsigned int * const adown_idx_ref, int * const maxi_ref, int * const mini_ref, double * const max_ref, double * const min_ref, double * const bar_ref, int * const j_ref, const double * const high, const double * const low, double * const adown, double * const aup, const int period, const double scale, int trail, int i)
{
  unsigned int adown_idx = *adown_idx_ref;
  int maxi = *maxi_ref;
  int mini = *mini_ref;
  double max = *max_ref;
  double min = *min_ref;
  double bar = *bar_ref;
  int j = *j_ref;
  bar = high[i];
  if (maxi < trail)
  {
    maxi = trail;
    max = high[maxi];
    j = trail;
    while ((++j) <= i)
    {
      bar = high[j];
      if (bar >= max)
      {
        max = bar;
        maxi = j;
      }
    }

  }
  else
    if (bar >= max)
  {
    maxi = i;
    max = bar;
  }
  bar = low[i];
  if (mini < trail)
  {
    mini = trail;
    min = low[mini];
    j = trail;
    while ((++j) <= i)
    {
      bar = low[j];
      if (bar <= min)
      {
        min = bar;
        mini = j;
      }
    }

  }
  else
    if (bar <= min)
  {
    mini = i;
    min = bar;
  }
  adown[adown_idx] = (((double) period) - (i - mini)) * scale;
  adown_idx += 1;
  *(aup++) = (((double) period) - (i - maxi)) * scale;
  *adown_idx_ref = adown_idx;
  *maxi_ref = maxi;
  *mini_ref = mini;
  *max_ref = max;
  *min_ref = min;
  *bar_ref = bar;
  *j_ref = j;
}


/*** DEPENDENCIES:
***/


int ti_aroon(int size, const double * const *inputs, const double *options, double * const *outputs)
{
  const double *high = inputs[0];
  const double *low = inputs[1];
  double *adown = outputs[0];
  unsigned int adown_idx = 0;
  double *aup = outputs[1];
  const int period = (int) options[0];
  if (period < 1)
  {
    return 1;
  }
  if (size <= ti_aroon_start(options))
  {
    return 0;
  }
  const double scale = 100.0 / period;
  int trail = 0;
  int maxi = -1;
  int mini = -1;
  double max = high[0];
  double min = low[0];
  double bar;
  int i;
  int j;
  for (i = period; i < size; i += 1, trail += 1)
  {
    helper_ti_aroon_1(&adown_idx, &maxi, &mini, &max, &min, &bar, &j, high, low, adown, aup, period, scale, trail, i);
  }

  assert(((&adown[adown_idx]) - outputs[0]) == (size - ti_aroon_start(options)));
  assert((aup - outputs[1]) == (size - ti_aroon_start(options)));
  return 0;
}


/*** DEPENDENCIES:
int ti_aroon_start(const double *options)
{
  return (int) options[0];
}


----------------------------
void helper_ti_aroon_1(unsigned int * const adown_idx_ref, int * const maxi_ref, int * const mini_ref, double * const max_ref, double * const min_ref, double * const bar_ref, int * const j_ref, const double * const high, const double * const low, double * const adown, double * const aup, const int period, const double scale, int trail, int i)
{
  unsigned int adown_idx = *adown_idx_ref;
  int maxi = *maxi_ref;
  int mini = *mini_ref;
  double max = *max_ref;
  double min = *min_ref;
  double bar = *bar_ref;
  int j = *j_ref;
  bar = high[i];
  if (maxi < trail)
  {
    maxi = trail;
    max = high[maxi];
    j = trail;
    while ((++j) <= i)
    {
      bar = high[j];
      if (bar >= max)
      {
        max = bar;
        maxi = j;
      }
    }

  }
  else
    if (bar >= max)
  {
    maxi = i;
    max = bar;
  }
  bar = low[i];
  if (mini < trail)
  {
    mini = trail;
    min = low[mini];
    j = trail;
    while ((++j) <= i)
    {
      bar = low[j];
      if (bar <= min)
      {
        min = bar;
        mini = j;
      }
    }

  }
  else
    if (bar <= min)
  {
    mini = i;
    min = bar;
  }
  adown[adown_idx] = (((double) period) - (i - mini)) * scale;
  adown_idx += 1;
  *(aup++) = (((double) period) - (i - maxi)) * scale;
  *adown_idx_ref = adown_idx;
  *maxi_ref = maxi;
  *mini_ref = mini;
  *max_ref = max;
  *min_ref = min;
  *bar_ref = bar;
  *j_ref = j;
}


----------------------------
***/

File: artifacts/tulipindicators/translate/src/aroon.rs
use crate::*;
use std::assert;

pub fn ti_aroon_start(options: Option<&[f64]>) -> i32 {
    // Use pattern matching to handle the Option safely
    match options {
        Some(opts) => {
            // Access the first element, default to 0 if slice is empty
            opts.first().map_or(0, |&val| val as i32)
        }
        None => 0, // Return 0 if options is None (NULL in C)
    }
}
pub fn ti_aroon(size: i32, inputs: Option<&[&[f64]]>, options: Option<&[f64]>, outputs: Option<&mut [&mut [f64]]>) -> i32 {
    // Check for None inputs (equivalent to NULL checks in C)
    if inputs.is_none() || options.is_none() || outputs.is_none() {
        return 1;
    }

    let inputs = inputs.unwrap();
    let options = options.unwrap();
    let outputs = outputs.unwrap();

    // Check if we have enough inputs and outputs
    if inputs.len() < 2 || outputs.len() < 2 {
        return 1;
    }

    let high = inputs[0];
    let low = inputs[1];
    
    // Split the outputs slice to get mutable references to both elements
    let (first, rest) = outputs.split_at_mut(1);
    let adown = &mut first[0];
    let aup = &mut rest[0];

    let period = options[0] as i32;
    if period < 1 {
        return 1;
    }

    let start = ti_aroon_start(Some(options));
    if size <= start {
        return 0;
    }

    let scale = 100.0 / period as f64;
    let mut trail = 0;
    let mut maxi = -1;
    let mut mini = -1;
    let mut max = high[0];
    let mut min = low[0];
    let mut bar;
    let mut adown_idx = 0;
    let mut aup_idx = 0;

    for i in period..size {
        bar = high[i as usize];
        if maxi < trail {
            maxi = trail;
            max = high[maxi as usize];
            let mut j = trail;
            while j <= i {
                bar = high[j as usize];
                if bar >= max {
                    max = bar;
                    maxi = j;
                }
                j += 1;
            }
        } else if bar >= max {
            maxi = i;
            max = bar;
        }

        bar = low[i as usize];
        if mini < trail {
            mini = trail;
            min = low[mini as usize];
            let mut j = trail;
            while j <= i {
                bar = low[j as usize];
                if bar <= min {
                    min = bar;
                    mini = j;
                }
                j += 1;
            }
        } else if bar <= min {
            mini = i;
            min = bar;
        }

        adown[adown_idx] = ((period - (i - mini)) as f64) * scale;
        adown_idx += 1;
        aup[aup_idx] = ((period - (i - maxi)) as f64) * scale;
        aup_idx += 1;
        trail += 1;
    }

    assert!((adown_idx as i32) == (size - start));
    assert!((aup_idx as i32) == (size - start));
    0
}
